sky/provision/kubernetes/instance.py

```diff
@@ -238,19 +244,33 @@ def _evaluate_timeout() -> bool:
         return time.time() - start_time < timeout
 
     while _evaluate_timeout():
-        all_pods_scheduled = True
-        for node in new_nodes:
-            # Iterate over each pod to check their status
-            pod = kubernetes.core_api(context).read_namespaced_pod(
-                node.metadata.name, namespace)
+        # Get all pods in a single API call using the cluster name label
+        # which all pods in new_nodes should share
+        cluster_name = new_nodes[0].metadata.labels[TAG_SKYPILOT_CLUSTER_NAME]
+        pods = kubernetes.core_api(context).list_namespaced_pod(
+            namespace,
+            label_selector=f'{TAG_SKYPILOT_CLUSTER_NAME}={cluster_name}').items
+
+        if len(pods) != len(pod_names):
+            logger.info('Retrying waiting for pods: '
+                        f'Expected {len(pod_names)} pods, got {len(pods)}')
+            time.sleep(0.5)
+            continue
+
+        # Filter to only the pods we're waiting for
+        pods = [pod for pod in pods if pod.metadata.name in pod_names]
```

Comment:
We already check that `len(pods) != len(pod_names)` above - so if this is filtering anything out then we don't actually have all the pods we are looking for, and we don't check that.
To be fair, it's pretty reasonable for it to be UB to have some random pod with the cluster name tag incorrectly set, but in that case shouldn't we just remove this filter? Otherwise could do a set comparison or something to check that the set of pods we got is exactly the set we expect.
Also, whatever behavior we follow here should be consistent with `_wait_for_pods_to_run`.

Reply:
good catch - updated to set equivalence checks.